pkg/storage/stores/shipper/downloads/table_manager.go

```diff
@@ -153,31 +153,27 @@ func (tm *TableManager) query(ctx context.Context, tableName string, queries []c
 }
 
 func (tm *TableManager) getOrCreateTable(tableName string) (*Table, error) {
-	// if table is already there, use it.
 	tm.tablesMtx.RLock()
+	defer tm.tablesMtx.RUnlock()
+
+	// if table is already there, use it.
 	table, ok := tm.tables[tableName]
-	tm.tablesMtx.RUnlock()
+	if ok {
+		return table, nil
+	}
 
-	if !ok {
-		tm.tablesMtx.Lock()
-		// check if some other competing goroutine got the lock before us and created the table, use it if so.
-		table, ok = tm.tables[tableName]
-		if !ok {
-			// table not found, creating one.
-			level.Info(util_log.Logger).Log("msg", fmt.Sprintf("downloading all files for table %s", tableName))
-
-			tablePath := filepath.Join(tm.cfg.CacheDir, tableName)
-			err := chunk_util.EnsureDirectory(tablePath)
-			if err != nil {
-				return nil, err
-			}
+	// table not found, creating one.
+	level.Info(util_log.Logger).Log("msg", fmt.Sprintf("downloading all files for table %s", tableName))
 
-			table = NewTable(tableName, filepath.Join(tm.cfg.CacheDir, tableName), tm.indexStorageClient, tm.boltIndexClient, tm.metrics)
-			tm.tables[tableName] = table
-		}
-		tm.tablesMtx.Unlock()
+	tablePath := filepath.Join(tm.cfg.CacheDir, tableName)
+	err := chunk_util.EnsureDirectory(tablePath)
+	if err != nil {
+		return nil, err
 	}
 
+	table = NewTable(tableName, filepath.Join(tm.cfg.CacheDir, tableName), tm.indexStorageClient, tm.boltIndexClient, tm.metrics)
+	tm.tables[tableName] = table
```

Comment:
don't you need to write lock here for this ?

Reply:
No, I missed that we're using `RLock` at the top. I've changed it to `Lock`